### PR TITLE
Enabled '_' in the field/column name in input type. 

### PR DESCRIPTION
**Pull request by zeahmed**
This PR fixes issue #318

### CODE REVIEW

**Comment by zeahmed** [comment]
Thanks @codemzs for heads up on this. I am going to add @TomFinley in the review. 

If column name can not accept '_' and name can be taken from field name within input type then it is restricting user to use C# variable naming convention.

File: src/Microsoft.ML/Data/TextLoader.cs [hunk 1]
```diff
@@ -84,8 +84,8 @@ public sealed partial class TextLoader
                         $"Valid characters are 0-9, *, - and ~");
 
                 var name = mappingAttr.Name ?? field.Name;
-                if (name.Any(c => !Char.IsLetterOrDigit(c)))
-                    throw Contracts.Except($"{name} is not alphanumeric.");
+                if (name.Any(c => !Char.IsLetterOrDigit(c) && c != '_'))
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML/Data/TextLoader.cs]: >!Char.IsLetterOrDigit(c) && c != '_') [](start = 34, length = 37)

I would rather use CSharpProvider.IsValidIdentifier

File: src/Microsoft.ML/Data/TextLoader.cs [hunk 2]
```diff
@@ -84,8 +85,8 @@ public sealed partial class TextLoader
                         $"Valid characters are 0-9, *, - and ~");
 
                 var name = mappingAttr.Name ?? field.Name;
-                if (name.Any(c => !Char.IsLetterOrDigit(c)))
-                    throw Contracts.Except($"{name} is not alphanumeric.");
+                if (!new CSharpCodeProvider().IsValidIdentifier(name))
+                    throw Contracts.Except($"{name} is not a valid identifier.");
```
Comment (TomFinley) [inline src/Microsoft.ML/Data/TextLoader.cs]: High level question -- I *believe* this is now using the new-ish API for v0.2 that utilizes the properties directly, and does not attempt to work through the goofy command line parsing that, IIRC, made this check necessary. So why do we still have it *at all*?

**Comment by TomFinley** [comment]
Hello @zeahmed , thank you for taking this on. High level question -- I *believe* this is now using the new-ish API for v0.2 that utilizes the properties directly, and does not attempt to work through the goofy command line parsing that, IIRC, made this check necessary. So why do we still have it *at all*? I think @codemzs made that transition #142 to use this, so perhaps he would know best.

If this code is being retained purely for my sake, please remove it. I would not say this is by design. It was by design back when we were using the command line parser so we had to worry about things like spaces and colons and everything screwing up the parsing... but since that poor design was improved in #142, as far I know this no longer applies. People can name their columns whatever they like more or less.

**Comment by mairaw** [comment]
If you keep the current behavior, it would be good if the error message is improved.

**Pull request merged**